src/main/scala/org/ergoplatform/nodeView/wallet/persistence/WalletRegistry.scala

```diff
@@ -128,20 +212,70 @@ final class WalletRegistry(store: LDBVersionedStore)(ws: WalletSettings) extends
   /**
     * Transits used boxes to a spent state or simply deletes them depending on a settings.
     */
-  private[persistence] def processHistoricalBoxes(spentBoxes: Seq[(ModifierId, TrackedBox)],
-                                                  spendingHeight: Int): RegistryOp[Unit] = {
+  private[persistence] def processHistoricalBoxes(bag: KeyValuePairsBag,
+                                                  spentBoxes: Seq[(ModifierId, TrackedBox)],
+                                                  spendingHeight: Int): KeyValuePairsBag = {
     if (keepHistory) {
-      updateBoxes(spentBoxes.map(_._2.box.id)) { tb =>
+      val outSpent: Seq[TrackedBox] = spentBoxes.flatMap { case (_, tb) =>
+        getBox(tb.box.id).orElse {
+          bag.toInsert.find(_._1.sameElements(key(tb))).flatMap { case (_, tbBytes) =>
+            TrackedBoxSerializer.parseBytesTry(tbBytes).toOption
+          } match {
+            case s@Some(_) => s
+            case None =>
+              log.warn(s"Output spent hasn't found in the wallet: ${Algos.encode(tb.box.id)}, " +
+                s"could be okay if it was created before wallet init")
+              None
+          }
+        }: Option[TrackedBox]
+      }
+
+      val updatedBoxes = outSpent.map { tb =>
         val spendingTxIdOpt = spentBoxes
-          .find { case (_, x) => encodedBoxId(x.box.id) == encodedBoxId(tb.box.id) }
+          .find { case (_, x) => x.box.id.sameElements(tb.box.id) }
           .map(_._1)
         tb.copy(spendingHeightOpt = Some(spendingHeight), spendingTxIdOpt = spendingTxIdOpt)
       }
+
+      val bagBeforePut = removeBoxes(bag, spentBoxes.map(_._2))
+      putBoxes(bagBeforePut, updatedBoxes)
     } else {
-      removeBoxes(spentBoxes.map(_._2.box.id))
+      removeBoxes(bag, spentBoxes.map(_._2))
     }
   }
 
+  /**
+    * Remove association between an application and a box
+    * @param appId application identifier
+    * @param boxId box identifier
+    */
+  def removeApp(appId: ApplicationId, boxId: BoxId): Try[Unit] = {
+    getBox(boxId) match {
+      case Some(tb) =>
+        (if (tb.applicationStatuses.size == 1) {
```

Comment:
In this case, if box is associated with just one application, it should be removed from the registly. Otherwise, only application <-> box reference to be removed.